### ISSUE TITLE
flux-start -s1 should not exec broker

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#834]
There are two code paths in `flux-start`, one in which `flux-start` provides a PMI environment for brokers that it spawns (size > 1), and one in which it simply calls `execve()` on the broker and gets out of the way (size = 1 or unspecified).

In the get-out-of-the-way path, it is assumed that the broker can sort itself out with PMI.   If launched under an RM like SLURM or Flux, the RM supplies the PMI bootstrap environment.  If launched standalone, the PMI "singleton" personality takes over.

Problem: these code paths supply different options to the broker, and this can cause confusion.  For example on @morrone's desktop, `flux start -s2` worked fine but `flux start -s1` failed with

```
lt-flux-broker: tcp://::1:*: No such device
```

in the former case, the broker was being spawned with `--shared-ipc-namespace` causing it to use an `ipc://` endpoint for its "child" socket.  In the latter case that option was omitted and it used a `tcp://` endpoint which failed due localhost mapping to the ipv6 `::1` address.

Perhaps `flux-start` should only get out of the way when the `-s,--size` option is completely omitted and not equate missing size with `--size=1`?

**Comment by trws** [comment]
This may also be related to an issue I've been running into with salloc.   Using `srun -n 1 -c 20 -N 1 flux start` works fine, but `salloc -N 1 -n 1 -c 20 flux start` fails with an error.

**Comment by garlick** [comment]
@trws please post that error if you get a chance.

**Comment by trws** [comment]
```
$ PYTHONPATH='' salloc -N 1 -n 1 -c 20 ./src/cmd/flux start ls
salloc: Granted job allocation 2806810
lt-flux-broker: PMI_KVS_Commit: operation failed
lt-flux-broker: bootstrap failed
salloc: Relinquishing job allocation 2806810
$ PYTHONPATH='' srun -N 1 -n 1 -c 20 ./src/cmd/flux start ls
config
config.log
config.status
doc
etc
flux-core-0.6.0-62-g9140754
flux-core-0.6.0-62-g9140754.tar.gz
install
libtool
lwj_dump.out
Makefile
print_args
src
t
```

**Comment by morrone** [comment]
> This may also be related to an issue I've been running into with salloc. Using srun -n 1 -c 20 -N 1 flux start works fine, but salloc -N 1 -n 1 -c 20 flux start fails with an error.

What were you hoping would happen with these commands?

In the srun case, you are asking for both a job allocation and a job step launch in the same command.  The job is allocated, and then a job step is launch remotely in the allocated node to run "flux start".  Since it is running as a slurm job step, presumably flux has full access to the PMI services offered by slurm and things would work normally.

In the salloc case, you are only asking for an allocation, and running the "flux start" command _locally_.  It does not have any access to PMI information (I am guessing this is why you see the "lt-flux-broker: PMI_KVS_Commit" error.  I forget off the top of my head what environment variables are and are not available after just an salloc.  But perhaps even more importantly "flux start" is not being executed on any of the nodes that were allocated.  So anything run under that flux instance would not be running on the nodes allocated by slurm either.  If you want "flux start" to run in the allocation, and you want to use salloc, the command should look more like this:

  salloc -N1 srun -n 1 -c 20 flux start

In this example, srun is _only_ doing the jobstep launch, since it learns from the environment that there is already an allocation.

I'm not clear on what the "-c 20" is doing for you with a "-n" of only 1 either.

**Comment by trws** [comment]
On 11 Jan 2017, at 16:37, Christopher J. Morrone wrote:

> In the salloc case, you are only asking for an allocation, and running 
> the "flux start" command _locally_.

Thanks for the clarification, clearly I've misunderstood salloc in the 
past.  Regardless, there's a bug here in that flux is looking for PMI 
information that doesn't exist due to the environment.

> I'm not clear on what the "-c 20" is doing for you with a "-n" of only 
> 1 either.

It's just insisting that I want 20 cores for that one task.

**Comment by chu11** [comment]
I was poking at this issue a little bit just to understand it.  One thing I didn't get, assuming ```::1``` is a properly configured loopback IP, shouldn't ```flux start -s1``` have still worked?  I tried hard coding "127.0.0.1" and its IPv6 mapped equivalents into the broker and that worked.  But ```::1``` and some variants of it didn't.  I will admit I'm not super IPv6 knowledgeable, perhaps there's some subtlety I'm missing.

**Comment by chu11** [comment]
self assigning as I think I get what's going on.  Hardest part will probably be updating the testsuite, as some tests use ```-s1``` and some don't, could be subtle behavior differences.

**Comment by garlick** [comment]
If there's an IPv6 issue let's open a new bug on it.   Here I was proposing a simpler change to `flux-start`:

> Perhaps flux-start should only get out of the way when the -s,--size option is completely omitted and not equate missing size with --size=1?

**Comment by chu11** [comment]
@garlick Understood, just want to make sure that this fix isn't to solve ```lt-flux-broker: tcp://::1:*: No such device```.  That ```tcp://::1:*``` not working may also be a bug.

**Comment by morrone** [comment]
I am going to suggest that the  "start_session" be moved to a completely separate flux subcommand...perhaps "localstart".  Jim explained that originally brokers were just started by hand in flux, but reviewers suggested that approach was confusing.  At the time, _flux start_ only did the start_session() part, and mostly for testing purposes I think.  But per reviewer feedback exec_broker() was added to _flux start_.

I think it would be best to now leave _flux start_ to just do the common user case, and move the original start_session() bits into its own command.

Better still might be to remove the start_session() code entirely from the _flux_ command.  We could instead have a "localpmi" launcher command that is mainly intended for testing.  In this way of thinking, testing would be run like this:

  localpmi -n 10 flux start

_localpmi_ would just set up pmi services and locally fork-exec N versions of the command following on the command line.  This way in testing we are always executing the exec_broker() code path that will be the common code path for users.

I'd be willing to work on this.  I'll wait until I get back on Tuesday and if you haven't gotten to it, @chu11, I'll steal it then. :)

**Comment by garlick** [comment]
See #643 where this was discussed previously.  I don't feel too strongly either way.

**Comment by garlick** [comment]
It's worth pointing out that it's the PMI code that we need in test (to get size > 1) but also that the other code path is practically a no-op so there isn't really a coverage issue there.

**Comment by chu11** [comment]
@morrone I like the ```localpmi -n 10 flux start``` idea.  I went ahead and fixed this specific issue in #951.  Perhaps we should make #952 into the bigger "re-arch" issue as such a change will hit many many things.  (Sorry, didn't see that this conversation thread had continued until after making #952).
